subsys/bluetooth/audio/bap_base.c

```diff
@@ -0,0 +1,562 @@
+
+				return NULL;
+			}
+
+			ltv_len = net_buf_simple_pull_u8(&net_buf);
+			if (net_buf.len < ltv_len) {
+				LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+				return NULL;
+			}
+			net_buf_simple_pull_mem(&net_buf, ltv_len);
+		}
+	}
+
+	return base;
+}
+
+int bt_bap_base_get_pres_delay(const struct bt_bap_base *base)
+{
+	struct net_buf_simple net_buf;
+	uint32_t pd;
+
+	CHECKIF(base == NULL) {
+		LOG_DBG("base is NULL");
+
+		return -EINVAL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)base, sizeof(pd));
+	pd = base_pull_pd(&net_buf);
+
+	return (int)pd; /* PD is 24-bit so it fits in an int */
+}
+
+int bt_bap_base_get_subgroup_count(const struct bt_bap_base *base)
+{
+	struct net_buf_simple net_buf;
+	uint8_t subgroup_count;
+
+	CHECKIF(base == NULL) {
+		LOG_DBG("base is NULL");
+
+		return -EINVAL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)base, BASE_MAX_SIZE);
+	base_pull_pd(&net_buf);
+	subgroup_count = net_buf_simple_pull_u8(&net_buf);
+
+	return (int)subgroup_count; /* subgroup_count is 8-bit so it fits in an int */
+}
+
+int bt_bap_base_foreach_subgroup(const struct bt_bap_base *base,
+				 bool (*func)(const struct bt_bap_base_subgroup *data,
+					      void *user_data),
+				 void *user_data)
+{
+	struct bt_bap_base_subgroup *subgroup;
+	struct net_buf_simple net_buf;
+	uint8_t subgroup_count;
+
+	CHECKIF(base == NULL) {
+		LOG_DBG("base is NULL");
+
+		return -EINVAL;
+	}
+
+	CHECKIF(func == NULL) {
+		LOG_DBG("func is NULL");
+
+		return -EINVAL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)base, BASE_MAX_SIZE);
+	base_pull_pd(&net_buf);
+	subgroup_count = net_buf_simple_pull_u8(&net_buf);
+
+	for (uint8_t i = 0U; i < subgroup_count; i++) {
+		subgroup = (struct bt_bap_base_subgroup *)net_buf.data;
+		if (!func(subgroup, user_data)) {
+			LOG_DBG("user stopped parsing");
+
+			return -ECANCELED;
+		}
+
+		/* Parse subgroup data to get next subgroup pointer */
+		if (subgroup_count > 1) { /* Only parse data if it isn't the last one */
+			uint8_t bis_count;
+
+			bis_count = base_pull_bis_count(&net_buf);
+			base_pull_codec_id(&net_buf, NULL);
+
+			/* Codec config */
+			base_pull_ltv(&net_buf, NULL);
+
+			/* meta */
+			base_pull_ltv(&net_buf, NULL);
+
+			for (uint8_t j = 0U; j < bis_count; j++) {
+				net_buf_simple_pull_u8(&net_buf); /* index */
+
+				/* Codec config */
+				base_pull_ltv(&net_buf, NULL);
+			}
+		}
+	}
+
+	return 0;
+}
+
+int bt_bap_base_get_subgroup_codec_id(const struct bt_bap_base_subgroup *subgroup,
+				      struct bt_bap_base_codec_id *codec_id)
+{
+	struct net_buf_simple net_buf;
+
+	CHECKIF(subgroup == NULL) {
+		LOG_DBG("subgroup is NULL");
+
+		return -EINVAL;
+	}
+
+	CHECKIF(codec_id == NULL) {
+		LOG_DBG("codec_id is NULL");
+
+		return -EINVAL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)subgroup, BASE_SUBGROUP_MAX_SIZE);
+	base_pull_bis_count(&net_buf);
+	base_pull_codec_id(&net_buf, codec_id);
```

Comment:
Would it be possible to move this to helper function to avoid duplicated code?

Reply:
Hmm, the checks are not the same, and the allocation of the `net_buf` would still need to be there, so I guess you are asking about moving 

```c
	net_buf_simple_init_with_data(&net_buf, (void *)subgroup, BASE_SUBGROUP_MAX_SIZE);
	base_pull_bis_count(&net_buf);
	base_pull_codec_id(&net_buf, codec_id);
```

into a helper function? 

It is something that we can do as something like 

```c
void base_init_net_buf_and_pull_subgroup_bis_and_codec_id(struct net_buf_simple *buf, struct bt_bap_base_codec_id *codec_id)
{
	net_buf_simple_init_with_data(&net_buf, (void *)subgroup, BASE_SUBGROUP_MAX_SIZE);
	base_pull_bis_count(&net_buf);
	base_pull_codec_id(&net_buf, codec_id);
}
```

It gets quite specific and it's also pretty hard to make a good name for it since it will do 3 things :D 

Is that what you requested?
Any suggestions on name?

Reply:
Right, let's leave it as it is.